lib/projects/localDev/LocalDevLogger.ts

```diff
@@ -0,0 +1,249 @@
+import { getAccountId } from '@hubspot/local-dev-lib/config';
+import { getConfigDefaultAccount } from '@hubspot/local-dev-lib/config';
+import { logger } from '@hubspot/local-dev-lib/logger';
+import chalk from 'chalk';
+
+import { uiLogger } from '../../ui/logger';
+import {
+  uiLink,
+  uiBetaTag,
+  uiLine,
+  UI_COLORS,
+  uiAccountDescription,
+  uiCommandReference,
+} from '../../ui';
+import { lib } from '../../../lang/en';
+import { LocalDevState } from '../../../types/LocalDev';
+import { getProjectDetailUrl } from '../../projects/urls';
+import SpinniesManager from '../../ui/SpinniesManager';
+
+class LocalDevLogger {
+  private state: LocalDevState;
+  private mostRecentUploadWarning: string | null;
+  private uploadWarnings: string[];
+
+  constructor(state: LocalDevState) {
+    this.state = state;
+    this.mostRecentUploadWarning = null;
+    this.uploadWarnings = [];
+  }
+
+  private logUploadInstructions(): void {
+    uiLogger.log('');
+    uiLogger.log(lib.LocalDevManager.uploadWarning.instructionsHeader);
+
+    uiLogger.log(lib.LocalDevManager.uploadWarning.stopDev);
+    if (this.state.isGithubLinked) {
+      uiLogger.log(lib.LocalDevManager.uploadWarning.pushToGithub);
+    } else {
+      uiLogger.log(
+        lib.LocalDevManager.uploadWarning.runUpload(this.getUploadCommand())
+      );
+    }
+    uiLogger.log(lib.LocalDevManager.uploadWarning.restartDev);
+  }
+
+  getUploadCommand(): string {
+    const currentDefaultAccount = getConfigDefaultAccount() || undefined;
+
+    return this.state.targetProjectAccountId !==
+      getAccountId(currentDefaultAccount)
+      ? uiCommandReference(
+          `hs project upload --account=${this.state.targetProjectAccountId}`
+        )
+      : uiCommandReference('hs project upload');
+  }
+
+  uploadWarning(): void {
+    // At the moment, there is only one additional warning. We may need to do this in a
+    // more robust way in the future
+    const additionalWarnings = this.uploadWarnings.join('\n\n');
+    const warning = `${lib.LocalDevManager.uploadWarning.defaultWarning} ${additionalWarnings}`;
+
+    // Avoid logging the warning to the console if it is currently the most
+    // recently logged warning. We do not want to spam the console with the same message.
+    if (warning !== this.mostRecentUploadWarning) {
+      uiLogger.log('');
+      uiLogger.warn(warning);
+      this.logUploadInstructions();
+
+      this.mostRecentUploadWarning = warning;
+    }
+  }
+
+  addUploadWarning(warning: string): void {
+    this.uploadWarnings.push(warning);
+  }
+
+  missingComponentsWarning(components: string[]): void {
+    const warning = lib.LocalDevManager.uploadWarning.missingComponents(
+      components.join(', ')
+    );
+
+    if (warning !== this.mostRecentUploadWarning) {
+      uiLogger.log('');
+      uiLogger.warn(warning);
+      this.logUploadInstructions();
+      this.mostRecentUploadWarning = warning;
+    }
+  }
+
+  fileChangeError(e: unknown): void {
```

Comment:
You could replace the next four functions with a single function like this:

```ts
  private handleError(
    e: unknown,
    translationFn: (message: string) => string
  ): void {
    if (this.state.debug) {
      logger.error(e);
    }
    uiLogger.error(translationFn(e instanceof Error ? e.message : ''));
  }
  ```

Reply:
Good call!